### PR TITLE
add Reduced class with tests for work with array_reduce in elegant way

### PR DESCRIPTION
**Pull request by otis22**
What do you think about this class in this library?

### CODE REVIEW

**Comment by otis22** [comment]
I will fix coverage

File: src/Reduced.php [hunk 1]
```diff
@@ -0,0 +1,55 @@
+<?php
+
+declare(strict_types=1);
+
+namespace ElegantBro\Arrayee;
+
+use ElegantBro\Interfaces\Arrayee;
+use Exception;
+use function array_merge;
```
Comment (stepanets) [inline src/Reduced.php]: it should be `array_reduce`

File: src/Reduced.php [hunk 2]
```diff
@@ -0,0 +1,55 @@
+<?php
+
+declare(strict_types=1);
+
+namespace ElegantBro\Arrayee;
```
Comment (stepanets) [inline src/Reduced.php]: I think we need sub-namespace, something like `ElegantBro\Arrayee\Aggregation`

File: src/Reduced.php [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+     * Reduced constructor.
+     * @param Arrayee $arrayee
+     * @param callable $callback
+     * @param Arrayee|null $initial
+     */
+    public function __construct(Arrayee $arrayee, callable $callback, ?Arrayee $initial = null)
+    {
+        $this->arrayee = $arrayee;
+        $this->callback = $callback;
+        $this->initial = $initial ?? new Just([]);
```
Comment (tigokr) [inline src/Reduced.php]: I have two points about this part: you shouldn't take this assert in the constructor and shouldn't use nullable arguments

try to create a secondary static constructor

**Review by tigokr — COMMENTED** [review]
it's great to see new contributors here =) nice work!

File: src/Reduced.php [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+use Exception;
+use function array_merge;
+
+final class Reduced implements Arrayee
+{
+    /**
+     * @var Arrayee
+     */
+    private $arrayee;
+    /**
```
Comment (stepanets) [inline src/Reduced.php]: empty line between properties required

File: src/Reduced.php (hunk 2 above)
Comment (otis22) [reply src/Reduced.php]: I will try fix it in next commit

File: src/Reduced.php (hunk 3 above)
Comment (stepanets) [reply src/Reduced.php]: Yeah, the primary constructor has no nullable arguments. And the secondary constructor will look like `initialEmpty(Arrayee $arrayee, callable $callback)`

File: src/Reduced.php [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    private $callback;
+    /**
+     * @var ?Arrayee
+     */
+    private $initial;
+
+    /**
+     * Reduced constructor.
+     * @param Arrayee $arrayee
+     * @param callable $callback
```
Comment (stepanets) [inline src/Reduced.php]: the description about `$callback` signature will be helpful `function(array $carry, $item): array`

Comment (otis22) [reply src/Reduced.php]: @stepanets i don't have knowledge about this hint

I mixed from this article https://phpstan.org/writing-php-code/phpdoc-types

callable(int $foo, string $bar): void (accepts an integer and a string, doesn’t return anything; parameter names are optional and insignificant)

And your comment

```php
    /**
     * Reduced constructor.
     * @param Arrayee $arrayee
     * @param callable(array, mixed): array $callback function(array $carry, $item): array
     * @param Arrayee $initial
     */
```
What do you think about?

File: src/Aggregation/Reduced.php [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+    /**
+     * @var Arrayee
+     */
+    private $initial;
+
+    /**
+     * Reduced constructor.
+     * @param Arrayee $arrayee
+     * @param callable(array, mixed): array $callback function(array $carry, $item): array
```
Comment (stepanets) [inline src/Aggregation/Reduced.php]: `callable(array, mixed): array` is not valid annotation type.

`@param callable $callback Function that reduces items to array <code>function(array $carry, $item): array</code>`

File: src/Aggregation/Reduced.php [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    public function __construct(Arrayee $arrayee, callable $callback, Arrayee $initial)
+    {
+        $this->arrayee = $arrayee;
+        $this->callback = $callback;
+        $this->initial = $initial;
+    }
+
+    /**
+     * @param Arrayee $arrayee
+     * @param callable(array, mixed): array $callback function(array $carry, $item): array
```
Comment (stepanets) [inline src/Aggregation/Reduced.php]: the same about type, and by convenience secondary constructors placed above the primary

Comment (otis22) [reply src/Aggregation/Reduced.php]: @stepanets I've fixed it both.

**Review by stepanets — APPROVED** [review]

**Pull request merged**